In [11]:
#Exercise 1: Exploratory Data Analysis

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from google.colab import files

uploaded = files.upload() ##upload the file

df = pd.read_csv("dataset_heart.csv")


print(df.shape) ##peek at the data
df.head()
df.info()
df.describe()


Saving dataset_heart.csv to dataset_heart (3).csv
(270, 14)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 270 entries, 0 to 269
Data columns (total 14 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   age                                   270 non-null    int64  
 1   sex                                   270 non-null    int64  
 2   chest pain type                       270 non-null    int64  
 3   resting blood pressure                270 non-null    int64  
 4   serum cholestoral                     270 non-null    int64  
 5   fasting blood sugar                   270 non-null    int64  
 6   resting electrocardiographic results  270 non-null    int64  
 7   max heart rate                        270 non-null    int64  
 8   exercise induced angina               270 non-null    int64  
 9   oldpeak                               270 non-null    float64
 10  ST segment                

age        sex   chest pain type  resting blood pressure  \
count  270.000000  270.000000       270.000000              270.000000   
mean    54.433333    0.677778         3.174074              131.344444   
std      9.109067    0.468195         0.950090               17.861608   
min     29.000000    0.000000         1.000000               94.000000   
25%     48.000000    0.000000         3.000000              120.000000   
50%     55.000000    1.000000         3.000000              130.000000   
75%     61.000000    1.000000         4.000000              140.000000   
max     77.000000    1.000000         4.000000              200.000000   

       serum cholestoral  fasting blood sugar  \
count         270.000000           270.000000   
mean          249.659259             0.148148   
std            51.686237             0.355906   
min           126.000000             0.000000   
25%           213.000000             0.000000   
50%           245.000000             0.000000   
75%           280.000000             0.000000   
max           564.000000             1.000000   

       resting electrocardiographic results  max heart rate  \
count                            270.000000      270.000000   
mean                               1.022222      149.677778   
std                                0.997891       23.165717   
min                                0.000000       71.000000   
25%                                0.000000      133.000000   
50%                                2.000000      153.500000   
75%                                2.000000      166.000000   
max                                2.000000      202.000000   

       exercise induced angina    oldpeak  ST segment  major vessels  \
count               270.000000  270.00000  270.000000     270.000000   
mean                  0.329630    1.05000    1.585185       0.670370   
std                   0.470952    1.14521    0.614390       0.943896   
min                   0.000000    0.00000    1.000000       0.000000   
25%                   0.000000    0.00000    1.000000       0.000000   
50%                   0.000000    0.80000    2.000000       0.000000   
75%                   1.000000    1.60000    2.000000       1.000000   
max                   1.000000    6.20000    3.000000       3.000000   

             thal  heart disease  
count  270.000000     270.000000  
mean     4.696296       1.444444  
std      1.940659       0.497827  
min      3.000000       1.000000  
25%      3.000000       1.000000  
50%      3.000000       1.000000  
75%      7.000000       2.000000  
max      7.000000       2.000000

In [ ]:
df.head()

age  sex   chest pain type  resting blood pressure  serum cholestoral  \
0   70     1                4                     130                322   
1   67     0                3                     115                564   
2   57     1                2                     124                261   
3   64     1                4                     128                263   
4   74     0                2                     120                269   

   fasting blood sugar  resting electrocardiographic results  max heart rate  \
0                    0                                     2             109   
1                    0                                     2             160   
2                    0                                     0             141   
3                    0                                     0             105   
4                    0                                     2             121   

   exercise induced angina  oldpeak  ST segment  major vessels  thal  \
0                        0      2.4           2              3     3   
1                        0      1.6           2              0     7   
2                        0      0.3           1              0     7   
3                        1      0.2           2              1     7   
4                        1      0.2           1              1     3   

   heart disease  
0              2  
1              1  
2              2  
3              1  
4              1

In [12]:
class_names = {1: 'No Heart Disease', 2: 'Has Heart Disease'}
print(df['heart disease'].value_counts().rename(index=class_names))

heart disease
No Heart Disease     150
Has Heart Disease    120
Name: count, dtype: int64


In [13]:
from sklearn.model_selection import train_test_split

#remove target column
y = df['heart disease']
X = df.drop('heart disease', axis=1)


X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.20, random_state=42, stratify=y
)

print(X_train.shape, X_test.shape)

#There are 270 rows with 216 training examples each with 13 features, 54 test rows with 13 features each.

(216, 13) (54, 13)


In [14]:
#Exercise 2 - Logistic Regression without Grid Search

from sklearn.linear_model import LogisticRegression

# Increase max_iter to help with convergence
model = LogisticRegression(max_iter=10000)

# we use .ravel() to transform the y_train data into a 1D array because LogisticRegression() expects the target variable (in this case, y_train) to be a 1D array rather than a column vector or a 2D array.

model.fit(X_train, y_train.values.ravel())

pred = model.predict(X_test)

In [15]:

from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

print("Accuracy:", accuracy_score(y_test, pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, pred))
print("Classification Report:\n", classification_report(y_test, pred))

Accuracy: 0.8518518518518519
Confusion Matrix:
 [[24  6]
 [ 2 22]]
Classification Report:
               precision    recall  f1-score   support

           1       0.92      0.80      0.86        30
           2       0.79      0.92      0.85        24

    accuracy                           0.85        54
   macro avg       0.85      0.86      0.85        54
weighted avg       0.86      0.85      0.85        54



#Evaluation of Model:

## The model achieved an accuracy of 85%, indicating strong overall performance.For example, it  identified 92% of patients with heart disease. However, there were 6 fall positives, and 2 false negatives. Performacne was balanced and there was no bias.

In [16]:
#Exercise 3 - Logistic Regression with Grid Search

# Necessary imports
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
import numpy as np
from sklearn.datasets import make_classification


X, y = make_classification(
    n_samples=1000, n_features=20, n_informative=10, n_classes=2, random_state=42)

# Creating the hyperparameter grid
c_space = np.logspace(-5, 8, 15)
param_grid = {'C': c_space}

# Instantiating logistic regression classifier
logreg = LogisticRegression()

# Instantiating the GridSearchCV object
logreg_cv = GridSearchCV(logreg, param_grid, cv=5)

# Assuming X and y are your feature matrix and target variable
# Fit the GridSearchCV object to the data
logreg_cv.fit(X, y)

# Print the tuned parameters and score
print("Tuned Logistic Regression Parameters: {}".format(logreg_cv.best_params_))
print("Best score is {}".format(logreg_cv.best_score_))
## cross-validated accuracy of 85.3% with C = 0.0061 and the lower C slightly improves performance.

Tuned Logistic Regression Parameters: {'C': np.float64(0.006105402296585327)}
Best score is 0.853


In [17]:
# Exercise 4: Create an SVM classifier with a linear kernel
from sklearn import svm
clf = svm.SVC(kernel='linear')

# Train the classifier on the training data
clf.fit(X_train, y_train)

# Make predictions
Y_pred = clf.predict(X_test)

In [18]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

print("Accuracy:", accuracy_score(y_test, Y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, Y_pred))
print("Classification Report:\n", classification_report(y_test, Y_pred))

##Evaluation: 85%  accuracy with balanced performance.

Accuracy: 0.8518518518518519
Confusion Matrix:
 [[25  5]
 [ 3 21]]
Classification Report:
               precision    recall  f1-score   support

           1       0.89      0.83      0.86        30
           2       0.81      0.88      0.84        24

    accuracy                           0.85        54
   macro avg       0.85      0.85      0.85        54
weighted avg       0.86      0.85      0.85        54



In [19]:
#Exercise 5: Exercise 5 : SVM with Grid Search
from sklearn import svm
from sklearn.model_selection import GridSearchCV
import numpy as np

# Create the parameter grid

param_grid = {
    'C': [0.1, 1, 10],
    'kernel': ['linear', 'rbf'],
    'gamma': ['scale']
}

# Instantiate the SVM classifier
svc = svm.SVC()

# Instantiate the GridSearchCV object
svc_cv = GridSearchCV(svc, param_grid, cv=5)

# Fit the GridSearchCV object to your data
svc_cv.fit(X_train, y_train.values.ravel())

# Print the tuned parameters and best score
print("Tuned SVM Parameters: {}".format(svc_cv.best_params_))
print("Best Cross-Validation Score: {:.3f}".format(svc_cv.best_score_))

##The SVM model tuned with GridSearchCV achieved a cross-validated accuracy of 83.8% with the best parameters C = 1, kernel = linear, and gamma = scale, confirming that a linear boundary best fits the dataset.

Tuned SVM Parameters: {'C': 1, 'gamma': 'scale', 'kernel': 'linear'}
Best Cross-Validation Score: 0.838


In [23]:
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import numpy as np

# Make sure labels are 0 and 1 for XGBoost
y_train_xgb = y_train.replace({1:0, 2:1}) if set(np.unique(y_train)) == {1,2} else y_train
y_test_xgb  = y_test.replace({1:0, 2:1})  if set(np.unique(y_test))  == {1,2} else y_test

# Manually set hyperparameters
model = XGBClassifier(
    learning_rate=0.1,
    n_estimators=200,
    max_depth=3,
    min_child_weight=2,
    subsample=0.8,
    colsample_bytree=0.8,
    gamma=0.0,
    reg_lambda=1.0,
    reg_alpha=0.0,
    random_state=42,
    eval_metric='logloss'
)

# Train and evaluate
model.fit(X_train, y_train_xgb)
y_pred = model.predict(X_test)

print(f"Accuracy: {accuracy_score(y_test_xgb, y_pred):.3f}")
print("Confusion Matrix:\n", confusion_matrix(y_test_xgb, y_pred))
print("Classification Report:\n", classification_report(y_test_xgb, y_pred))

#I picked learning_rate=0.1 and n_estimators=200 because a smaller learning rate makes training more stable, while a higher number of trees helps the model learn enough patterns. I used a max_depth of 3 to prevent overfitting since this dataset isn’t very large. Setting subsample=0.8 and colsample_bytree=0.8 helps the model generalize better by using only a portion of the data and features for each tree.

Accuracy: 0.796
Confusion Matrix:
 [[23  7]
 [ 4 20]]
Classification Report:
               precision    recall  f1-score   support

           0       0.85      0.77      0.81        30
           1       0.74      0.83      0.78        24

    accuracy                           0.80        54
   macro avg       0.80      0.80      0.80        54
weighted avg       0.80      0.80      0.80        54



In [ ]:
##My manual XGBoost model achieved solid performance with these parameters. The model trained efficiently, avoided overfitting, etc

In [24]:
#Exercise 7: from xgboost import XGBClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import numpy as np

# Make sure labels are 0 and 1 for XGBoost
y_train_xgb = y_train.replace({1:0, 2:1}) if set(np.unique(y_train)) == {1,2} else y_train
y_test_xgb  = y_test.replace({1:0, 2:1})  if set(np.unique(y_test))  == {1,2} else y_test

# Define the base model
xgb = XGBClassifier(
    use_label_encoder=False,
    eval_metric='logloss',
    random_state=42
)

# Define the parameter grid for GridSearchCV
param_grid = {
    'learning_rate': [0.01, 0.1, 0.2],
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 4, 5],
    'subsample': [0.8, 1],
    'colsample_bytree': [0.8, 1]
}

# Set up GridSearchCV
xgb_cv = GridSearchCV(
    estimator=xgb,
    param_grid=param_grid,
    cv=5,
    scoring='accuracy',
    verbose=1,
    n_jobs=-1
)

# Fit the grid search to the training data
xgb_cv.fit(X_train, y_train_xgb)

# Print the best parameters and CV score
print("Best Parameters:", xgb_cv.best_params_)
print(f"Best Cross-Validation Accuracy: {xgb_cv.best_score_:.3f}")

# Evaluate the best model on the test set
best_xgb = xgb_cv.best_estimator_
y_pred = best_xgb.predict(X_test)

print(f"Test Accuracy: {accuracy_score(y_test_xgb, y_pred):.3f}")
print("Confusion Matrix:\n", confusion_matrix(y_test_xgb, y_pred))
print("Classification Report:\n", classification_report(y_test_xgb, y_pred))


Fitting 5 folds for each of 108 candidates, totalling 540 fits


/usr/local/lib/python3.12/dist-packages/xgboost/training.py:183: UserWarning: [12:44:43] WARNING: /workspace/src/learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


Best Parameters: {'colsample_bytree': 0.8, 'learning_rate': 0.1, 'max_depth': 4, 'n_estimators': 200, 'subsample': 0.8}
Best Cross-Validation Accuracy: 0.829
Test Accuracy: 0.796
Confusion Matrix:
 [[23  7]
 [ 4 20]]
Classification Report:
               precision    recall  f1-score   support

           0       0.85      0.77      0.81        30
           1       0.74      0.83      0.78        24

    accuracy                           0.80        54
   macro avg       0.80      0.80      0.80        54
weighted avg       0.80      0.80      0.80        54



#GridSearchCV picked a moderately regularized XGBoost (depth 4, 200 trees, lr 0.1) that scored 0.829 CV and 0.796 test accuracy, trailing my linear baselines—evidence that a linear decision boundary fits this dataset best.